api/auth/auth.cpp

```diff
@@ -232,47 +234,187 @@ error::Error FetchJWTToken(
+				PostPendingActions(ex_auth_data);
+			}
+		});
+	if (err != error::NoError) {
+		// A sync DBus error.
+		mlog::Error("Failed to request new token fetching: " + err.String());
+		token_fetch_in_progress_ = false;
+		ExpectedAuthData ex_auth_data = expected::unexpected(err);
+		PostPendingActions(ex_auth_data);
+		if (opt_action) {
+			(*opt_action)(ex_auth_data);
+		}
+		return err;
+	}
+	// else everything went OK
+
+	token_fetch_in_progress_ = true;
+
+	// Make sure the action (if any) is called once the token is
+	// obtained.
+	if (opt_action) {
+		pending_actions_.push_back(*opt_action);
+	}
+
+	// Make sure we don't wait for the token forever.
+	auth_timeout_timer_.AsyncWait(auth_timeout_, [this](error::Error err) {
+		if (err.code == make_error_condition(errc::operation_canceled)) {
+			return;
+		} else if (err == error::NoError) {
+			mlog::Warning("Timed-out waiting for a new token");
+			token_fetch_in_progress_ = false;
+			ExpectedAuthData ex_auth_data = expected::unexpected(
+				MakeError(AuthenticationError, "Timed-out waiting for a new token"));
+			PostPendingActions(ex_auth_data);
+		} else {
+			// should never happen
+			assert(false);
+			mlog::Error("Authentication timer error: " + err.String());
+
+			// In case it did happen, run the stacked up actions or things may
+			// got stuck.
+			ExpectedAuthData ex_auth_data = expected::unexpected(err);
+			PostPendingActions(ex_auth_data);
+		}
+	});
+	return error::NoError;
+}
+
 error::Error Authenticator::WithToken(AuthenticatedAction action) {
-	if (token_) {
-		string token = *token_;
-		action(ExpectedToken(token));
+	if (!watching_token_signal_) {
+		auto err = StartWatchingTokenSignal();
+		if (err != error::NoError) {
+			// Should never fail. We rely on the signal heavily so let's fail
+			// hard if it does.
+			action(expected::unexpected(err));
+			return err;
+		}
+	}
+
+	if (token_ && server_url_) {
+		AuthData auth_data {*server_url_, *token_};
+		action(ExpectedAuthData(auth_data));
 		return error::NoError;
 	}
 	// else => no token
-	if (auth_in_progress_) {
+
+	if (token_fetch_in_progress_) {
+		// Already waiting for a new token, just make sure the action is called
+		// once it arrives (or once the wait times out).
 		pending_actions_.push_back(action);
 		return error::NoError;
 	}
 	// else => should fetch the token, cache it and call all pending actions
-	auth_in_progress_ = true;
-
-	error::Error err = FetchJWTToken(
-		client_,
-		server_url_,
-		private_key_path_,
-		device_identity_script_path_,
-		[this, action](APIResponse resp) {
-			if (resp) {
-				token_ = resp.value();
+	// Try to get token from mender-auth
+	auto err = dbus_client_.CallMethod<dbus::ExpectedStringPair>(
+		"io.mender.AuthenticationManager",
+		"/io/mender/AuthenticationManager",
+		"io.mender.Authentication1",
+		"GetJwtToken",
+		[this, action](dbus::ExpectedStringPair ex_auth_dbus_data) {
+			token_fetch_in_progress_ = false;
```

Comment:
Nothing prevents `Authenticator::WithToken()` to be called multiple times in a row while the DBus round-trip of the `GetJwtToken()` method call happens.

Reply:
I'm a bit lost. So if you have this sequence of events:

1. Calls DBus.FetchJwtToken
2. While waiting, calls DBus.GetJwtToken
3. GetJwtToken returns immediately, with an empty token.
4. The `token_fetch_in_progress_ = false` statement executes, signalling that fetching is over, but in fact, the FetchJwtToken call is still running.

Isn't that wrong? Or am I missing something?

Reply:
I believe what you are missing is that in case `FetchJwtToken()` has been called (and the signal hasn't delivered a new token yet), `token_fetch_in_progress_` would be `true` and thus `GetJwtToken()` wouldn't be called, the given `action` would only be added to the pending actions.

So this particular statement really is only for what I described above -- multiple fast consecutive calls of `WithToken()` in the window when the DBus communication happens. If the `GetJwtToken()` method's reply arrives before the next `WithToken()`, there will either be a token, or `FetchJwtToken()` will be called. In both cases a following `WithToken()` call won't call `GetJwtToken()` (it will either add a pending action or use the cached token).

Or is it me missing something?

Reply:
Ah ok, now I see what you mean. From the perspective of mender-update (which is where all this is called from), both operations are fetching operations. I was thinking only of FetchJwtToken as a fetching operation. So if either is running, it doesn't matter which, subsequent operations are queued up, so it will never run both in parallel, correct?

Then I agree! 👍

Reply:
Yes, exactly! `FetchJwtToken()` in ***mender-auth*** is a different story. Here in the code **relying** on *mender-auth* it's really just token fetching being in progress (due to async nature of our code an DBus). That's why I renamed the boolean to `token_fetch_in_progress_` (used to be `auth_in_progress_`).